service/share/share_test.go

```diff
@@ -135,3 +142,822 @@ func BenchmarkService_GetSharesByNamespace(b *testing.B) {
+				if bytes.Equal(tc.namespaces[i], namespace) {
+					msgsInNamespace = append(msgsInNamespace, tc.messages[i])
+				}
+			}
+
+			namespacedShares, _ := b.Data.ComputeShares()
+
+			// test round trip using only encoding, without IPLD
+			{
+				myShares := make([][]byte, 0)
+				for _, sh := range namespacedShares.RawShares() {
+					if bytes.Equal(namespace, sh[:8]) {
+						myShares = append(myShares, sh)
+					}
+				}
+				msgs, err := core.ParseMsgs(myShares)
+				require.NoError(t, err)
+				assert.Len(t, msgs.MessagesList, len(msgsInNamespace))
+				for i := range msgs.MessagesList {
+					assert.Equal(t, msgsInNamespace[i], msgs.MessagesList[i].Data)
+				}
+			}
+
+			// test full round trip - with IPLD + decoding shares
+			{
+				extSquare, err := ipld.PutData(ctx, namespacedShares.RawShares(), store)
+				require.NoError(t, err)
+
+				dah := da.NewDataAvailabilityHeader(extSquare)
+				shares, err := serv.GetSharesByNamespace(ctx, &dah, namespace)
+				require.NoError(t, err)
+				require.NotEmpty(t, shares)
+
+				rawShares := make([][]byte, len(shares))
+				for i, share := range shares {
+					rawShares[i] = share.Data()
+				}
+
+				msgs, err := core.ParseMsgs(rawShares)
+				require.NoError(t, err)
+				assert.Len(t, msgs.MessagesList, len(msgsInNamespace))
+				for i := range msgs.MessagesList {
+					assert.Equal(t, namespace, []byte(msgs.MessagesList[i].NamespaceID))
+					assert.Equal(t, msgsInNamespace[i], msgs.MessagesList[i].Data)
+				}
+			}
+		})
+	}
+}
+
+// this is a sample block from devnet-2 which originally showed the issue with share ordering
+// nolint:lll
+var sampleBlock = []byte(`
+    {
+      "header": {
+        "version": {
+          "block": 11
+        },
+        "chain_id": "devnet-2",
+        "height": 825320,
+        "time": "2022-04-14T07:52:33.890556391Z",
+        "last_block_id": {
+          "hash": "7FA631D8E8DA005132615B5BB9DD715E66ABE18DB3BF4A891242CCF090D629D1",
+          "parts": {
+            "total": 1,
+            "hash": "9C2705130B4D8A9A9FAD64D462944EABC281B5AC36ADBFA6E3A48A7F347DD8C1"
+          }
+        },
+        "last_commit_hash": "00AAFEB5461158D198967E750E108367EE6D0739E69C0AC2F4D102E0F87D2595",
+        "data_hash": "2AE354BAA0A6C75EDAAB7AA9D6FAAF873FD6FF3E0884DD479306EC6C1AAFD796",
+        "validators_hash": "A76865C38C998C200C6056F8B1D08CB4009683B1110ACE85D95B9889A47650ED",
+        "next_validators_hash": "A76865C38C998C200C6056F8B1D08CB4009683B1110ACE85D95B9889A47650ED",
+        "consensus_hash": "048091BC7DDC283F77BFBF91D73C44DA58C3DF8A9CBC867405D8B7F3DAADA22F",
+        "app_hash": "E2DD3E32D66B330B16BCAB389DA17C2B6BC1B082B2B58ED0A90E8197E36A59A6",
+        "last_results_hash": "E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855",
+        "evidence_hash": "E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855",
+        "proposer_address": "2CB2349B8516D69839FE49892B1AF6909D7361A7"
+      },
```

Comment:
As all this contains specifics/internal of the tendermint block + commits, can we add a comment for future maintainers how to handle this test-vector in case there was a breaking change in tendermint (e.g. what if we switched to BLS, or, remove the partSetHeader etc).

Reply:
Both swamp, as well as e2e tests in the ephemeral cluster would make it unnecessary to hardcode test-vectors that contain and spill over implementation details of structs defined in other repos.

Reply:
I don't like the fact that we need to copy the JSON structure here for the test, as we should be ideally making a real block with an integration test, but we are blocked on the mocked app, so this way turns out to be the simplest one atm.